evm/utils/test_builder/test_builder.py

```diff
@@ -0,0 +1,218 @@
+from collections import (
+    defaultdict,
+)
+from functools import (
+    partial,
+)
+
+from evm.db.state import (
+    MainAccountStateDB,
+)
+
+from cytoolz import (
+    assoc_in,
+    curry,
+    merge,
+)
+from eth_utils import (
+    apply_formatters_to_dict,
+    decode_hex,
+    encode_hex,
+    to_canonical_address,
+)
+
+from .normalization import (
+    normalize_environment,
+    normalize_state,
+    normalize_transaction,
+    normalize_transaction_group,
+    normalize_networks,
+)
+from .builder_utils import (
+    add_transaction_to_group,
+    calc_state_root,
+    get_version_from_git,
+    merge_nested,
+    wrap_in_list,
+)
+from .formatters import (
+    filled_formatter,
+)
+
+
+#
+# Defaults
+#
+
+DEFAULT_ENVIRONMENT = {
+    "currentCoinbase": to_canonical_address("0x2adc25665018aa1fe0e6bc666dac8fc2697ff9ba"),
+    "currentDifficulty": 131072,
+    "currentGasLimit": 1000000,
+    "currentNumber": 1,
+    "currentTimestamp": 1000,
+    "previousHash": decode_hex(
+        "0x5e20a0453cecd065ea59c37ac63e079ee08998b6045136a8ce6635c7912ec0b6"
+    ),
+}
+
+DEFAULT_TRANSACTION = {
+    "data": b"",
+    "gasLimit": 100000,
+    "gasPrice": 0,
+    "nonce": 0,
+    "secretKey": decode_hex("0x45a915e4d060149eb4365960e6a7a45f334393093061116b197e3240065ff2d8"),
+    "to": to_canonical_address("0x0f572e5295c57f15886f9b263e2f6d2d6c7b5ec6"),
+    "value": 0
+}
+
+ALL_NETWORKS = [
+    "Frontier",
+    "Homestead",
+    "EIP150",
+    "EIP158",
+    "Byzantium"
+]
+
+ACCOUNT_STATE_DB_CLASSES = {
+    "Frontier": MainAccountStateDB,
+    "Homestead": MainAccountStateDB,
+    "EIP150": MainAccountStateDB,
+    "EIP158": MainAccountStateDB,
+    "Byzantium": MainAccountStateDB,
+}
+assert all(network in ACCOUNT_STATE_DB_CLASSES for network in ALL_NETWORKS)
+
+FILLED_WITH_TEMPLATE = "py-evm-{version}"
+
+
+#
+# Filler Generation
+#
+
+def setup_filler(name, environment=None):
+    environment = merge(DEFAULT_ENVIRONMENT, normalize_environment(environment or {}))
+    return {name: {
+        "env": environment,
+    }}
+
+
+@curry
+def pre_state(pre_state, filler):
+    assert len(filler) == 1
+    test_name = next(iter(filler.keys()))
```

Comment:
These two lines are duplicated a few times.  Maybe write a small utility function  **or** change the signature to expect two arguments `pre_state(pre_state, test_name, state)`